app/gui/src/project-view/components/MarkdownEditor/markdown/trees.ts

```diff
@@ -0,0 +1,556 @@
+/** @file Lezer Tree operations for reading and writing inline formatting state. */
+import {
+  zeroFormatDepths,
+  type FormatDepths,
+  type FormatNode,
+  type NormalizedRange,
+  type SeminormalizedRange,
+  type TrimmedRange,
+} from '@/components/MarkdownEditor/markdown/types'
+import { type SyntaxNodeRef, type Tree, type TreeCursor } from '@lezer/common'
+import { identity } from '@vueuse/core'
+import { Range } from 'ydoc-shared/util/data/range'
+
+function reversed<T>(elements: ReadonlyArray<T>): T[] {
+  return elements.slice().reverse()
+}
+
+interface Side {
+  fromOrTo: 'from' | 'to'
+  toOrFrom: 'to' | 'from'
+  inside: 1 | -1
+  outside: -1 | 1
+  innermost: (a: number, b: number) => number
+  outermost: (a: number, b: number) => number
+  insideOrder: <T>(nodes: T[]) => T[]
+  outsideOrder: <T>(nodes: T[]) => T[]
+}
+const FROM: Side = {
+  fromOrTo: 'from',
+  toOrFrom: 'to',
+  inside: 1,
+  outside: -1,
+  innermost: Math.max,
+  outermost: Math.min,
+  insideOrder: identity,
+  outsideOrder: reversed,
+}
+const TO: Side = {
+  fromOrTo: 'to',
+  toOrFrom: 'from',
+  inside: -1,
+  outside: 1,
+  innermost: Math.min,
+  outermost: Math.max,
+  insideOrder: reversed,
+  outsideOrder: identity,
+}
+
+function sides<T>(f: (side: Side) => T): { from: T; to: T } {
+  return {
+    from: f(FROM),
+    to: f(TO),
+  }
+}
+
+const MARK_NODE: Readonly<Record<FormatNode, string>> = {
+  Emphasis: 'EmphasisMark',
+  StrongEmphasis: 'EmphasisMark',
+  Strikethrough: 'StrikethroughMark',
+}
+
+abstract class TreeRangeVisitor {
+  protected constructor(protected readonly range: Range) {}
+
+  protected abstract enter(node: SyntaxNodeRef): boolean | void
+  protected leave(_node: SyntaxNodeRef): boolean | void {}
+
+  abstract visit(tree: Tree): void
+}
+
+abstract class ExclusiveTreeVisitor extends TreeRangeVisitor {
+  visit(tree: Tree) {
+    tree.iterate({
+      from: this.range.from,
+      to: this.range.to,
+      enter: (node) => {
+        if (node.to === this.range.from || node.from === this.range.to) return false
+        return this.enter(node)
+      },
+      leave: this.leave.bind(this),
+    })
+  }
+}
+
+abstract class ContainedNodeVisitor extends TreeRangeVisitor {
+  private isFullyContained(node: SyntaxNodeRef): boolean {
+    return this.range.from <= node.from && node.from <= this.range.to
```

Comment:
if "fully", this should be:
```suggestion
  private isFullyContained(node: SyntaxNodeRef): boolean {
    return this.range.from <= node.from && node.to <= this.range.to
```

Reply:
It is effectively covered, a lot of tests were failing